docs/DafnyRef/UserGuide.md

```diff
@@ -1733,7 +1733,35 @@ Detailed description of the `dafny build` and `dafny run` commands and
 the `--input` option (needed when `dafny run` has more than one input file)
 is contained [in the section on command-line structure](#command-line).
 
-### 13.8.2. C\#
+### 13.8.2. Placeholder modules
+To enable easily customising runtime behavior across an entire Dafny program, Dafny has placeholder modules. Here follows an example:
+
+<!-- %check-run -->
+```dafny
+replaceable module Foo {
+  method Bar() returns (i: int) 
+    ensures i >= 2
+}
+
+method Main() {
+  var x := Foo.Bar();
+  print x;
+}
+// At this point, the program can be verified but not run.
+
+module ConcreteFoo replaces Foo {
+  method Bar() returns (i: int) {
+    return 3; // Main will print 3.
+  }
+}
+// ConcreteFoo can be swapped out for different replacements of Foo, to customize runtime behavior.
+```
+
+When replacing a placeholder module, the same rules apply as when refining an abstract module. However, unlike an abstract module, a placeholder module can be used as if it is a concrete module. When executing code, using for example `dafny run` or `dafny translate`, any program that contains a placeholder module must also contain a replacement of this placeholder. When using `dafny verify`, placeholder modules do not have to be replaced.
```

Comment:
I think I would love it if I could define a compilable replaceable module that has all its declaration bodies (e.g. a slow reference implementation that I can use to run my model and iterate on the design), and replaces module would actually swap it with e.g. a better implementation.

But that way, the replaceable module could be enough, it's just it would make it possible to make a dependency injection and allow for mocked testing for example.

How hard would it be/acceptable to let 100% defined replaceable modules be still compiled as usual if no replacement module is found?

Reply:
> How hard would it be/acceptable to let 100% defined replaceable modules be still compiled as usual if no replacement module is found?

For that we must enable module refinement to override bodied functions and methods.

Reply:
Oh, interesting. I thought replacement was meant to actually replace methods while keeping their signatures. To me that would be one of the biggest advantages of replaceable modules.